### ISSUE TITLE
Generalization & Enhancement of customized events, e.g. `CustomPeriodicEvent()`

### ISSUE THREAD

**Issue by DrStoop** [pytorch/ignite#648]
### Issue

The current `CustomPeriodicEvent` and other classes like `EarlyStopping`, `NanOutput` (more to come) share common features like condition checking, condition updating and firing events which could/should be inherited from a base class. Also the current `CustomPeriodicEvents` is limited to iteration/epoch-number-periodicity of `Events.ITERATION_COMPLETED` and `Events.EPOCH_COMPLETED`, therefore this should be extended to all possible use-cases.
Limits of current `CustomPeriodicEvent`:
1. no base/meta classes for feature sharing
2. event-fire-condition fixed to number of iterations and epochs
3. customized events of multiple custom events are not listed in a central dictionary, e.g. `Events` for overview
4. customized events have non-customizable standardized event names (e.g. `Events.ITERATIONS_10`) instead of e.g. `Events.SAMPLES_1000_COMPLETED`, `Events.EARLY_STOPPING`, `Events.NAN_OUTPUT` etc. 

### Solution

To define base/meta classes (no. 1. of above list) consider following hierarchical breakdown:
- The Events are all fired at certain conditions (e.g. certain number of iterations is reached, loss limit is surpassed etc.), some of them periodic some of them not. So one could talk about `BaseConditionalEvents`.
- For periodic events two events per iteration would be fired, i.e. `Events.*_STARTED` & `Events.*_COMPLETED`, whereas for more specific conditions, e.g. early stopping, only a single `Events.EARLY_STOPPING` is required. Therefore a `BaseConditionalSingleEvents` and a `BaseConditionalDoubleEvents` could inherit from `BaseConditionalEvents`.
- The conditions to be checked, how they are updated also depending on if an event was fired or not and when the conditions are reset vary strongly with use-cases and should be defined as `abstractmethod`s (no. 2.).
- Also to which events these methods are attached should be provided as arguments to the users. I propose initial event arguments  compatible to the suggested enhanced `Metric` class in  #643, with reasonable default values facilitating backward compatibility (`started_event`, `step_started_event`, `step_completed_event`, `completed_event`) (no. 2).
- Generally all parameter of the `engine`'s `State` should be stored to `engine.state`. So this would theoretically include the condition's parameter of all custom event together (e.g. current minimum loss) and the number of fired custom events. Nevertheless with multiple conditional events the parameter quickly spam `engine.state`. So I propose only saving the number of fired events for each conditional event to `engine.state` and keep the other parameter in the conditional event instance. Also for debugging it is easier to have all parameter in the instance of the custom event than in one of the `engine`s it is attached to.

To cope with the other limits listed above I suggest:
3. add all customized events as `<enum Events>` to `Events` so all existing events can be viewed from everywhere
4. provide argument `name_of_event` in `__init__()` to customize the event name... because the `<enum Events>` have the (slightly confusing) argument `event_name` I suggest the `name_of_event` for distinguishing... (or just change 'event_name' to 'event'? don't know)

### Code suggestion

I provide in the following 2 modules (+1 to make them run before PR):
* `base_conditional_events.py` including the base classes
* `step_periodic_events.py` including two applications of the base classes: the `__name__=='__main__'` sections provides an output of `Events` holding all custom events
* (`logger_mixin.py` including the logger mixin class just in case someone wants to test before the PR... logger would be an issue on its own and and possibly synced with PyTorch)

I'm always happy about feedback/improvements, so let me know...

#### `base_conditional_events.py`
```python
from abc import ABCMeta, abstractmethod
from aenum import extend_enum

from ignite.engine import Events, State
from ignite._six import with_metaclass


class BaseConditionalEvents(with_metaclass(ABCMeta, object)):
    """
    Base class for all repetitive event handlers triggered when pre-defined condition(s) are reached.
    Normally, this meta class may only be inherited by other meta/base classes (e.g.
    `BaseConditionalsingle/DobuleEvents`) but not directly by event handler classes

    Note: The fired events counter should be stored and updated in `engine.state` as `event_state_attr`.

    Args:
        name_of_event (str): prefix of the event's full name, with `num_period_steps` being the suffix
        started_event (<enum, 'Events'>): trigger event to initialize any firing of conditional events. Before
            this `started_event` no event is fired at all.
        step_completed_event (<enum, 'Events'>): trigger event for firing `self.conditional_completed_event`
        completed_event (<enum, 'Events'>): trigger event to terminate any further firing of conditional events.
            After this event no more events are fired at all.

    """

    def __init__(self, name_of_event, started_event=Events.STARTED, step_completed_event=Events.ITERATION_COMPLETED,
                 completed_event=Events.EPOCH_COMPLETED):
        # Assign parameters
        self.started_event = started_event
        self.step_completed_event = step_completed_event
        self.completed_event = completed_event
        # Attribute name of `self.step_completed_event`'s iteration counter in `engine.state`
        self.step_event_state_attr = self._extract_event_state_attr(name_of_event=None, suffix=None,
                                                                    event_name=self.step_completed_event)
        # Iteration counter for fired `self.conditional_completed_events` set in `engine.state`
        if not hasattr(self, 'conditional_state_attr'):
            self.conditional_state_attr = name_of_event
        self._initialize_all_conditional_events()
        # Controls time interval (defined by `self.started_event`, `self.completed_event) in which conditional event is fired.
        self._firing_enabled = False

    @property
    def name_of_event(self):
        """Tag for print outs or help for manual debugging."""
        return self.conditional_state_attr

    @staticmethod
    def _extract_event_state_attr(name_of_event=None, event_name=None, suffix=None):
        """
        Extracts name components from `event_state_attr`, `event_name` and `suffix` and combines all to
        `engine.state`-attribute of the step events which is assigned to `self.conditional_state_attr`.
        The prefix is either from `name_of_event` (if not `None`) or is extracted from `event_name.value`.
        """
        # Either cut of `_STARTED` or `_COMPLETED` from `step_event.value` to receive `self.step_event_state_attr`
        if name_of_event is None and event_name is None:
            raise TypeError('Either or both of `name_of_event` and `event_name` must not be `None`, '
                            'but both values are `None`.')
        if name_of_event is None:
            if event_name.value.split(sep='_')[-1].upper() in ['STARTED', 'COMPLETED']:
                name_of_event = '_'.join(event_name.value.split(sep='_')[:-1]).lower()
            # or use full `self..step_started_event
            else:
                name_of_event = event_name.value.lower()
        return '_'.join([name_of_event.lower(), str(suffix)]) if suffix is not None \
            else name_of_event

    @abstractmethod
    def _initialize_all_conditional_events(self):
        """
        Generate full name of conditional events whereas the number of events can vary for different
        event handlers (i.e. custom events). This method should normally be implemented in the next
        inheriting base/meta class, e.g. `BaseConditionalSingle/DoubleEvents`.
        """

    def _initialize_conditional_event(self, name_of_event):
        """
        Any custom event must be initialized equally, meaning:
            * standardized uppercase enum `value`s:
                * single event have arbitrary name `name_of_event`
                * double events have arbitrary prefix `name_of_event` and standard suffix `_STARTED` or `_COMPLETED`
            * standardized lowercase enum `name`s identical to their `value`s
            * custom event is extended to `Events`, so any custom event created will be listed there
            * custom event is extended to `State` with key `self.conditional_state_attr`, so all `conditional_state_attr`s
              are listete there
        """
        # Create conditional event by extending `Events`
        # Event existence in `Events` has to be checked and avoided because multiple conditional events with
        # same event names and values cloud be created during model training. This can mess things up, no good!!!
        if not hasattr(Events, name_of_event.upper()) and True:
            # Create event by extending `Events`
            extend_enum(Events, name_of_event.upper(), name_of_event.lower())
        else:
            raise RuntimeError('`{}` already exists in `{}` and therefore cannot be created. '
                  'Change current `name_of_event = {}` to avoid name collision.'.format(
                getattr(Events, name_of_event.upper()), Events, name_of_event.lower()))
        # Extend `State.event_to_attr` with conditional event
        State.event_to_attr[getattr(Events, name_of_event.upper())] = self.conditional_state_attr.lower()
        return getattr(Events, name_of_event.upper())

    @abstractmethod
    def _reset(self, engine):
        """
        Resets reset counters `conditional_state_attr`(s) in `engine.state` to initial value, e.g. like this:
            .. code-block:: python

                setattr(engine.state, self.conditional_event_state_attr, 1)

        Args:
            engine: the engine to which the `self` is attached to
        """
        pass

    @abstractmethod
    def _check_condition_to_fire_conditional_completed_event(self, engine) -> bool:
        """
        Checks conditions and returns `True` to fire `self.conditional_event_completed_event` in the current step
        and returns `False` to hold back the firing.

        Args:
            engine: the engine to which the `self` is attached to
        """
        pass

    @abstractmethod
    def _update_condition_if_fired_conditional_completed_event(self, engine):
        """
        Updates the fire-event-condition parameter(s) after firing `self.conditional_completed_event`.
        E.g.

        Args:
            engine: the engine to which the `self` is attached to
        """
        pass

    @abstractmethod
    def _update_condition_if_not_fired_conditional_completed_event(self, engine):
        """
        Updates the fire-event-condition parameter(s) after NOT firing `self.conditional_completed_event` or
        `self.conditional_event`.

        Args:
            engine: the engine to which the `self` is attached to
        """
        pass
    
    def _on_started(self, engine):
        self._firing_enabled = True
        self._reset(engine)

    def _on_step_completed(self, engine):
        """Fires `self.periodic_completed_event` when `self.n_period_steps` is reached. Therefore this method
        is called with each `self.step_completed_event`."""
        if self._check_condition_to_fire_conditional_completed_event(engine) and self._firing_enabled:
            engine.fire_event(event_name=self.conditional_completed_event)
            self._update_condition_if_fired_conditional_completed_event(engine)
        else:
            self._update_condition_if_not_fired_conditional_completed_event(engine)

    def _on_completed(self, engine):
        self._firing_enabled = False

    def attach(self, engine):
        # Add event handlers
        # avoid adding the same event handler multiple times:
        if not engine.has_event_handler(handler=self._on_started, event_name=self.started_event):
            engine.add_event_handler(handler=self._on_started, event_name=self.started_event)
        else:
            raise RuntimeError('{}` event handler has already been added to engine `{}` at event `{}`. You '
                               'should not add multiple times.'.format(self._on_started,  engine, self.started_event))
        # avoid adding the same event handler multiple times:
        if not engine.has_event_handler(handler=self._on_step_completed, event_name=self.step_completed_event):
            engine.add_event_handler(self.step_completed_event, self._on_step_completed)
        else:
            raise RuntimeError('{}` event handler has already been added to engine `{}` at event `{}`. You '
                               'should not add multiple times.'.format(self._on_started, engine, self.started_event))


class BaseConditionalSingleEvents(with_metaclass(ABCMeta, BaseConditionalEvents)):
    """
    Base class for conditional event generator which fire a single event `self.conditional_completed_event`
    at each completed step if the implemented conditions are reached.
    Args:
        name_of_event (str): prefix of the event's full name, with `num_period_steps` being the suffix
        started_event (<enum, 'Events'>): trigger event to initialize any firing of conditional events. Before
            this `started_event` no event is fired at all.
        step_completed_event (<enum, 'Events'>): trigger event for firing `self.conditional_completed_event`
        completed_event (<enum, 'Events'>): trigger event to terminate any further firing of conditional events.
                                            After this event no more events are fired at all.
    """

    def _initialize_all_conditional_events(self):
        """Initialize the single-event `{NAME_OF_EVENT}`."""
        self.conditional_completed_event = self._initialize_conditional_event(name_of_event=self.conditional_state_attr)

    def attach(self, engine):
        engine.register_events(self.conditional_completed_event)
        super(BaseConditionalSingleEvents, self).attach(engine)


class BaseConditionalDoubleEvents(with_metaclass(ABCMeta, BaseConditionalEvents)):
    """
    Base class for conditional event generator which fire a two events (double event) at each started
    and completed step, namely `self.conditional_started_event` and `self.conditional_completed_event`,
     if the implemented conditions are reached.
    Args:
        name_of_event (str): prefix of the event's full name, with `num_period_steps` being the suffix
        started_event (<enum, 'Events'>): trigger event to initialize any firing of conditional events. Before
            this `started_event` no event is fired at all.
        step_started_event (<enum, 'Events'>): trigger event for firing `self.conditional_started_event`
        step_completed_event (<enum, 'Events'>): trigger event for firing `self.conditional_completed_event`
        completed_event (<enum, 'Events'>): trigger event to terminate any further firing of conditional events.
                                            After this event no more events are fired at all.
    """
    def __init__(self, name_of_event, started_event=Events.STARTED, step_started_event=Events.ITERATION_STARTED,
                 step_completed_event=Events.ITERATION_COMPLETED, completed_event=Events.EPOCH_COMPLETED):
        self.step_started_event = step_started_event
        self.conditional_started_event = None  # value initialized in `self.intialize_conditional_events()`
        self.conditional_completed_event = None  # value initialized in `self.intialize_conditional_events()`
        super(BaseConditionalDoubleEvents, self).__init__(name_of_event, started_event,
                                                          step_completed_event, completed_event)

    def _initialize_all_conditional_events(self):
        """Initialize the double-event of `{NAME_OF_EVENT}_STARTED` and `{NAME_OF_EVENT}_COMPLETED`."""
        conditional_names_of_events = [suffix.format(self.conditional_state_attr.lower())
                                               for suffix in ["{}_started", "{}_completed"]]
        conditional_event_attrs = ['conditional_started_event', 'conditional_completed_event']
        for event_attr, name_of_event in zip(conditional_event_attrs, conditional_names_of_events):
            setattr(self, event_attr, self._initialize_conditional_event(name_of_event))

    @abstractmethod
    def _check_condition_to_fire_started(self, engine) -> bool:
        """
        Checks conditions and returns `True` to fire `self.conditional_started_event` in the current step
        and returns `False` if conditions are not met.

        Args:
            engine: the engine to which the `self` is attached to
        """
        pass

    @abstractmethod
    def _update_condition_if_fired_started(self, engine):
        """
        Updates the fire-event-condition parameter(s) after firing `self.conditional_started_event`.

        Args:
            engine: the engine to which the `self` is attached to
        """
        pass

    def _update_condition_if_not_fired_started(self, engine):
        """
        Updates the fire-event-condition parameter(s) after NOT firing `self.conditional_started_event`.

        Args:
            engine: the engine to which the `self` is attached to
        """
        pass

    def _on_step_started(self, engine):
        if self._check_condition_to_fire_started(engine) and self._firing_enabled:
            engine.fire_event(event_name=self.conditional_started_event)
            self._update_condition_if_fired_started(engine)
        else:
            self._update_condition_if_not_fired_started(engine)

    def attach(self, engine):
        engine.register_events(self.conditional_started_event, self.conditional_completed_event)
        # avoid adding the same event handler multiple times:
        if not engine.has_event_handler(handler=self._on_step_started, event_name=self.step_started_event):
            engine.add_event_handler(self.step_started_event, self._on_step_started)
        else:
            raise RuntimeError('{}` event handler has already been added to engine `{}` at event `{}`. '
                               'You should not add multiple times.'.format(self._on_step_started, engine,
                                                                           self.step_started_event))
        super(BaseConditionalDoubleEvents, self).attach(engine)
```

#### `step_periodic_events.py`
```python
from ignite.engine import Events, State
from ignite_contrib.handlers import BaseConditionalDoubleEvents, BaseConditionalSingleEvents, LoggerMixin


class StepPeriodicEvents(BaseConditionalSingleEvents):
    """
    Fires the custom event `self.conditional_completed_event` on every `num_period_steps`-th occurrence
    of `step_completed_event`.

    Args:
        num_period_steps (int): number of `step_started_events` to occur between two sequential events
                                `self.conditional_started_event` or `self.conditional_completed_event`
        name_of_event (str): prefix of the event's full name, with `num_period_steps` being the suffix
        started_event (<enum, 'Events'>): trigger event to initialize any firing of conditional events. Before
                                          this event no event is fired at all.
        step_started_event (<enum, 'Events'>): trigger event for firing `self.conditional_started_event`
        step_completed_event (<enum, 'Events'>): trigger event for firing `self.conditional_completed_event`
        completed_event (<enum, 'Events'>): trigger event to terminate any further firing of conditional events.
                                            After this event no more events are fired at all.
    """
    def __init__(self, n_period_steps, name_of_event=None, started_event=Events.STARTED,
                 step_completed_event=Events.ITERATION_COMPLETED, completed_event=Events.COMPLETED):
        self.n_period_steps = n_period_steps
        # `self.conditional_event_state_attr` is also returned by @property `self.name_of_event`
        self.conditional_event_state_attr = self._extract_event_state_attr(name_of_event=name_of_event,
                                                          suffix=self.n_period_steps,
                                                          event_name=step_completed_event)
        super(StepPeriodicEvents, self).__init__(name_of_event=self.conditional_event_state_attr,
                                                 started_event=started_event,
                                                 step_completed_event=step_completed_event,
                                                 completed_event=completed_event)

    def _reset(self, engine):
        # Initialization/reset value set to `1` instead of `0` representing the number of the current iteration
        setattr(engine.state, self.conditional_event_state_attr, 1)

    def _check_condition_to_fire_conditional_completed_event(self, engine):
        return getattr(engine.state, self.step_event_state_attr) % self.n_period_steps == 0

    def _update_condition_if_fired_conditional_completed_event(self, engine):
        """Updates only the fired-conditional-event-counter of `engine.state` but not the condition."""
        setattr(engine.state, self.conditional_event_state_attr, getattr(engine.state, self.conditional_event_state_attr) + 1)

    def _update_condition_if_not_fired_conditional_completed_event(self, engine):
        pass


class StepPeriodicDoubleEvents(BaseConditionalDoubleEvents):
    """
    Fires the custom event `self.conditional_started_event` at the be beginning and the custom event
    `self.conditional_completed_event` at the end of an interval. The interval is defined by the number
    'num_period_steps` of `step_completed_event`s that have to occur for triggering a fire-event.

    Args:
        num_period_steps (int): number of `step_started_events` to occur between two sequential events
                                `self.conditional_started_event` or `self.conditional_completed_event`
        name_of_event (str): prefix of the event's full name, with `num_period_steps` being the suffix
        started_event (<enum, 'Events'>): trigger event to initialize any firing of conditional events. Before
                                          this event no event is fired at all.
        step_started_event (<enum, 'Events'>): trigger event for firing `self.conditional_started_event`
        step_completed_event (<enum, 'Events'>): trigger event for firing `self.conditional_completed_event`
        completed_event (<enum, 'Events'>): trigger event to terminate any further firing of conditional events.
                                            After this event no more events are fired at all.
    """
    def __init__(self, n_period_steps, name_of_event=None, started_event=Events.STARTED,
                 step_started_event=Events.ITERATION_STARTED, step_completed_event=Events.ITERATION_COMPLETED,
                 completed_event=Events.COMPLETED):
        self.n_period_steps = n_period_steps
        # `self.conditional_event_state_attr` is also returned by @property `self.name_of_event`
        self.conditional_event_state_attr = self._extract_event_state_attr(name_of_event=name_of_event,
                                                                     suffix=self.n_period_steps,
                                                                     event_name=step_completed_event)
        super(StepPeriodicDoubleEvents, self).__init__(name_of_event=self.conditional_event_state_attr,
                                                       started_event=started_event,
                                                       step_started_event=step_started_event,
                                                       step_completed_event=step_completed_event,
                                                       completed_event=completed_event)

    def _reset(self, engine):
        # Initialization/reset value set to `1` instead of `0` representing the number of the current iteration
        setattr(engine.state, self.conditional_event_state_attr, 1)

    def _check_condition_to_fire_started(self, engine):
        return getattr(engine.state, self.step_event_state_attr) % self.n_period_steps == 1

    def _update_condition_if_fired_started(self, engine):
        """
        Updates only the fired-conditional-event-counter of `engine.state` but not the condition. Only the
        fired `stelf.conditional_started_event`s are counted.
        """
        setattr(engine.state, self.conditional_event_state_attr, getattr(engine.state, self.conditional_event_state_attr) + 1)

    def _update_condition_if_not_fired_started(self, engine):
        pass

    def _check_condition_to_fire_conditional_completed_event(self, engine):
        return getattr(engine.state, self.step_event_state_attr) % self.n_period_steps == 0

    def _update_condition_if_fired_conditional_completed_event(self, engine):
        pass

    def _update_condition_if_not_fired_conditional_completed_event(self, engine):
        pass


if __name__ == '__main__':

    custom_periodic_event = StepPeriodicDoubleEvents(n_period_steps=10)

    named_custom_periodic_event = StepPeriodicDoubleEvents(n_period_steps=10, name_of_event='ten_iterations')

    print('\nCustomized events appear in `Events` identified by idividual (if provided) or standard name:')
    for event_name, event in Events.__members__.items():
        print('{}: {}'.format(event_name, event))
```
Run this module on its own gives the `Events` holding all current/initiated custom events:
```python
Customized events appear in `Events` identified by idividual (if provided) or standard name:
EPOCH_STARTED: Events.EPOCH_STARTED
EPOCH_COMPLETED: Events.EPOCH_COMPLETED
STARTED: Events.STARTED
COMPLETED: Events.COMPLETED
ITERATION_STARTED: Events.ITERATION_STARTED
ITERATION_COMPLETED: Events.ITERATION_COMPLETED
EXCEPTION_RAISED: Events.EXCEPTION_RAISED
ITERATION_10_STARTED: Events.ITERATION_10_STARTED
ITERATION_10_COMPLETED: Events.ITERATION_10_COMPLETED
TEN_ITERATIONS_10_STARTED: Events.TEN_ITERATIONS_10_STARTED
TEN_ITERATIONS_10_COMPLETED: Events.TEN_ITERATIONS_10_COMPLETED
```

#### `logger_mixin.py` (not part of the issue, just for running this code)
```python
from abc import ABCMeta, abstractmethod, abstractproperty
from ignite._six import with_metaclass
import logging

# ======================================================================================================================
### GLOBAL LOGGER PARAMETERS ###
# ======================================================================================================================
# Change one of these parameters to change them globally
GLOBAL_LOGGING_LEVEL = logging.INFO
GLOBAL_HANDLER_CLASS = logging.StreamHandler
GLOBAL_LOGGING_FORMAT = '[%(levelname)s:%(name)s.%(funcName)s():%(lineno)s] %(message)s'
GLOBAL_MULTIPROCESSING_LOGGING_FORMAT = '[%(processName)s/%(levelname)s:%(name)s.%(funcName)s():%(lineno)s] %(message)s'


def update_formatter_of_handler_if_none(hdlr, fmt):
    if not hdlr.formatter:
        global_formatter = logging.Formatter(fmt)
        hdlr.setFormatter(global_formatter)
    return hdlr


def replace_handler(logger, hdlr):
    """
    Replaces the last handler for regular loggers and the first hander for multiprocessing loggers in
    `logger.handlers` with `the handler provided in argument`hdlr`.
    If `hdlr` has no formatter, then a formatter with `fmt` is added.

    Args:
        hdlr (logging.Handler): handler to replace the current one
    """
    if not isinstance(hdlr, logging.Handler):
        raise TypeError('Argument `hdlr` should be of type `logging.Handler`, but given {}.'.format(
            type(hdlr)))
    if 'multiprocessing' in logger.name:
        hdlr = update_formatter_of_handler_if_none(hdlr, fmt=GLOBAL_MULTIPROCESSING_LOGGING_FORMAT)
        # `multiprocessing` is limited to one handler secured by a lock against simultaneous process calls
        logging.getLogger(name='multiprocessing').handlers[0] = hdlr
    else:
        hdlr = update_formatter_of_handler_if_none(hdlr, fmt=GLOBAL_LOGGING_FORMAT)
        to_be_replaced_handler = logger.handlers[-1]
        if to_be_replaced_handler == hdlr:
            logger.warn('The logger`s handler is being repalced by itself provided by argument `repalcing_handler`.'
                        '`MixinLogger.get_logger()` needs to be called only once this the `hdlr` to'
                        'change the local handler permanently.')

        logger.removeHandler(hdlr=to_be_replaced_handler)
        logger.addHandler(hdlr=hdlr)


def get_logger(name):
    """
    Get logger with `name` from `logging` and set it to global parameters.

    Args:
        name (str): logger name
    :return:
    """
    # If logger already exists
    if name in logging.root.manager.loggerDict.keys():
        return logging.getLogger(name=name)
    # If logger does not exist yet and needs to be created
    logger = logging.getLogger(name=name)
    logger.setLevel(level=GLOBAL_LOGGING_LEVEL)
    formatter = logging.Formatter(GLOBAL_LOGGING_FORMAT)
    handler = GLOBAL_HANDLER_CLASS()
    handler.setFormatter(fmt=formatter)
    logger.addHandler(hdlr=handler)
    return logger


# Multiprocessing logger
# For logging.DEBUG level change level in train()
def get_multiprocessing_logger(name):
    # In the name `multiprocessing` is separated on purpose by `.` as `logging` build hierachy on this seperator
    multiprocessing_name = 'multiprocessing' + '.' + name
    # If mp_logger already exists
    if multiprocessing_name in logging.root.manager.loggerDict.keys():
        return logging.getLogger(name=multiprocessing_name)
    # If mp_logger does not exist yet and needs to be created
    mp_logger = logging.getLogger(name=multiprocessing_name)
    mp_logger.setLevel(level=GLOBAL_LOGGING_LEVEL)
    mp_formatter = logging.Formatter(fmt=GLOBAL_MULTIPROCESSING_LOGGING_FORMAT)
    mp_handler = GLOBAL_HANDLER_CLASS()
    mp_handler.setFormatter(fmt=mp_formatter)
    # Override the automatically generated standard-`multiprocessing`-handler with customized handler `mp_handler`
    if len(logging.getLogger(name='multiprocessing').handlers) == 0:
        logging.getLogger(name='multiprocessing').handlers.append(mp_handler)
    else:
        logging.getLogger(name='multiprocessing').handlers[0] = mp_handler
    return mp_logger


class BaseLoggerMixin(object):
    @abstractproperty
    def _logger(self):
        pass

    @abstractmethod
    def _logger_replace_handler(self, hdlr):
        pass

    @abstractmethod
    def _logger_reformat_all_handlers(self, format_string):
        pass


class LoggerMixin(BaseLoggerMixin):
    """
    Mixin to provide any class with a standardized logger `self._logger` property with standardized logger name.
    """
    @property
    def _logger(self):
        return get_logger(name=__file__ + "." + self.__class__.__name__.__str__())

    def _logger_replace_handler(self, hdlr):
        """
        Replaces the last (normally the only) handler in `logger.handlers` by `the handler provided in argument hdlr`.
        If `hdlr` has no formatter, then a formater with `GLOBAL_LOGGING_FORMAT` or
        `GLOBAL_MULTIPROCESSING_LOGGING_FORMAT` is added.

        Args:
             hdlr (logging.Handler): a instantiated handler to replace the current handler
         """
        replace_handler(logger=self._logger, hdlr=hdlr)

    def _logger_reformat_all_handlers(self, format_string):
        for hdlr in self._logger.handlers:
            hdlr.setFormatter(fmt=logging.Formatter(fmt=format_string))


class MultiprocessingLoggerMixin(BaseLoggerMixin):
    """
    Mixin to provide any class with a `mulitprocessing` standardized logger `self._logger` property with standardized
    logger name.
    """
    @property
    def _logger(self):
        return get_multiprocessing_logger(name=__file__ + "." + self.__class__.__name__.__str__())

    def _logger_replace_handler(self, hdlr):
       replace_handler(logger=self._logger, hdlr=hdlr)

    def _logger_reformat_all_handlers(self, format_string):
        logging.getLogger(name='multiprocessing').handlers[0].setFormatter(
            fmt=logging.Formatter(fmt=format_string))


if __name__ == '__main__':

    class DummyClassForLogging(LoggerMixin, object):
        pass


    class DummyClassForMultiprocessingLogging(MultiprocessingLoggerMixin, object):
        pass

    for dummy in [DummyClassForLogging(), DummyClassForMultiprocessingLogging()]:
        dummy._logger.critical('If you can read this critical log, it`s correct.')
        dummy._logger.error('If you can read this error log, it`s correct.')
        dummy._logger.info('This is a logger test. If this info log is displayed on stdout, it`s correct.')
        dummy._logger.warn('If you can read this warning log, it`s correct.')
        dummy._logger.debug('If you can read this debug log, something is wrong!')

        # Change the logger level of the single instance to debug it
        dummy._logger.setLevel(level=logging.DEBUG)
        dummy._logger.debug('If you can read this debug log NOW after changing the logger level, it`s correct.')

        # Change the logger handler of the single instance to debug it
        filelogger_handler = logging.FileHandler(filename='please_delete_me.log')
        dummy._logger_replace_handler(hdlr=filelogger_handler)
        dummy._logger.debug('If this debug log is logged to the file, it`s correct.')
        dummy._logger.info('If this info log is also logged to the file, it`s wrong!')
        dummy._logger_replace_handler(hdlr=filelogger_handler)
        dummy._logger.debug('If this is logged together with a WARNING to the file, it`s correct.')
        # Change the logger handler back again
        dummy._logger_replace_handler(hdlr=logging.StreamHandler())
        dummy._logger.info('If this info log is stdoutted again, it`s correct.')
        # Change the format
        changed_format = '!!!CHANGED FORAMT!!! [%(levelname)s:%(name)s.%(funcName)s():%(lineno)s] %(message)s'
        dummy._logger_reformat_all_handlers(format_string=changed_format)
        dummy._logger.info('If this info log message has changed format, it`s correct.')
```

**Comment by vfdev-5** [comment]
@DrStoop thanks for the issue and the propositions ! Please, correct me if I misunderstood something in your idea.
Mainly, I agree that we need to generalize and improve events system to be more customizable and such idea was recently proposed in https://github.com/pytorch/ignite/pull/624 
I think `CustomPeriodicEvent` is hard to use and we probably need to propose something light and easy to use directly with `Events`. Basic idea of #624 is too provide a synthetic sugar for 
```python
def event_filter(engine):
     # some logic
     return True # of False

@trainer.on(Events.EPOCH_STARTED, *args, **kwargs)
def user_handler(engine, *args, **kwargs):
    if event_filter(engine):
        # run handler
```
such that user could call event handler in any kind of pattern:
```
def random_event_filter(engine):
    return random.rand() < 0.5

trainer.add_event_handler(Events.EPOCH_STARTED(random_event_filter), handler)
trainer.add_event_handler(Events.EPOCH_STARTED(every(5)), handler)
trainer.add_event_handler(Events.EPOCH_STARTED(pattern_XX_XXX), handler)
```


> 4. customized events have non-customizable standardized event names (e.g. Events.ITERATIONS_10) instead of e.g. Events.EARLY_STOPPING, Events.NAN_OUTPUT

Yes, this is interesting !

**Comment by DrStoop** [comment]
Agree, we need more sugar! ...I guess I'll think about the sugar more like in #624 before proceedings with issues/PRs... may take a bit, haven't got the fastest brain ;)

**Comment by Bibonaut** [comment]
Currently, I am also investigating the code to get an understanding what can be realized with proposition from #624 and this proposal. But I also need some time to dig in :)
